src/Umbraco.Web/Routing/RedirectTrackingEventHandler.cs

```diff
@@ -89,62 +91,98 @@ protected override void ApplicationStarted(UmbracoApplicationBase umbracoApplica
+                    RequestCache.GetCacheItem(ContextKey2, () => true);
+                }
                 else
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey2);
+                {
+                    RequestCache.ClearCacheItem(ContextKey2);
+                }
             }
         }
 
         private static bool Moving
         {
-            get { return UmbracoContext.Current.HttpContext.Items[ContextKey1] != null; }
+            get { return RequestCache.GetCacheItem(ContextKey1) != null; }
             set
             {
                 if (value)
-                    UmbracoContext.Current.HttpContext.Items[ContextKey1] = true;
+                {
+                    //this forces true into the cache
+                    RequestCache.GetCacheItem(ContextKey1, () => true);
+                }
                 else
                 {
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey1);
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey2);
+                    RequestCache.ClearCacheItem(ContextKey1);
+                    RequestCache.ClearCacheItem(ContextKey2);
                 }
             }
         }
 
+        /// <summary>
+        /// Before the items are published, we need to get it's current URL before it changes
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
         private static void ContentService_Publishing(IPublishingStrategy sender, PublishEventArgs<IContent> args)
         {
             if (LockedEvents) return;
 
-            var contentCache = UmbracoContext.Current.ContentCache;
+            var contentCache = GetPublishedCache();
+            if (contentCache == null) return;
+
             foreach (var entity in args.PublishedEntities)
             {
-                var entityContent = contentCache.GetById(entity.Id);
-                if (entityContent == null) continue;
-                foreach (var x in entityContent.DescendantsOrSelf())
+                //TODO: This is horrible - we need to check if the url segment for this entity is changing in 
+                // order to determine if we need to make redirects for itself and all of it's descendents.
+                // The way this works right now (7.5.0) is that we re-lookup the entity that is currently being published which
+                // returns it's existing data in the db which we use to extract it's current segment. Then we compare that with
+                // the segment value returned from the current entity.
+                // In the future this will certainly cause some problems, to fix this we'd need to change the IUrlSegmentProvider
+                // to support being able to determine if a segment is going to change for an entity. See notes in IUrlSegmentProvider.
+                var oldEntity = ApplicationContext.Current.Services.ContentService.GetById(entity.Id);
```

Comment:
I have changed the way this detects if the URL is going to change based on the IUrlSegmentProviders. BUT, as you can see by the above notes, this is hacky and might break in future versions if something changes regarding our services/repository caches, etc... Plus it's just not ideal to go looking up an older version this way. Instead in 7.5+ we should make an additional interface for IUrlSegmentProvider for us to query if a segment will change depending on the dirty properties of the content item.